controllers/common/helper.go

```diff
@@ -698,3 +698,51 @@ func GetRecommendedCRVersion(
 	}
 	return nil, nil //nolint:nilnil
 }
+
+// HandleDBDeletion handles deletion of the resources related to the database Cluster
+// Returns true if the deletion is complete.
+func HandleDBDeletion(
+	ctx context.Context,
+	c client.Client,
+	database *everestv1alpha1.DatabaseCluster,
+	status everestv1alpha1.DatabaseClusterStatus,
+) (bool, error) {
+	if controllerutil.ContainsFinalizer(database, DBDeleteFinalizer) {
+		// updating the status to make the AppStateDeleting effective
+		database.Status = status
+		if err := c.Status().Update(ctx, database); err != nil {
+			return false, err
+		}
+		if done, err := checkDatabaseDeletion(ctx, c, database.GetName(), database.GetNamespace()); err != nil {
+			return false, err
+		} else if !done {
+			return false, nil
+		}
+		controllerutil.RemoveFinalizer(database, DBDeleteFinalizer)
+		return true, c.Update(ctx, database)
+	}
+	return true, nil
+}
+
+// Checks if the created for the DB resources are deleted.
+// Returns true if the deletion is complete.
+func checkDatabaseDeletion(
+	ctx context.Context,
+	c client.Client,
+	dbName, dbNs string,
+) (bool, error) {
+	backupList, err := ListDatabaseClusterBackups(ctx, c, dbName, dbNs)
+	if err != nil {
+		return false, err
+	}
+
+	restoresList, err := ListDatabaseClusterRestores(ctx, c, dbName, dbNs)
+	if err != nil {
+		return false, err
+	}
```

Comment:
I don't know if we should be checking the restores at all, since the deletion of restores is controlled using OwnerReferences.

Since the OwnerReference contains the `blockOwnerDeletion` flag, instead of explicitly checking for restores, we should perform a foreground deletion from the Everest API. That way, Kubernetes will guarentee the the DatabaseCluster stays as long as the restores are not deleted.

Reply:
Here's an example of a foreground deletion we do for backups: https://github.com/percona/everest/blob/7ecd61206826ff849839bd1bb25ce87107870556/api/database_cluster_backup.go#L105

Reply:
This will make sure that Kubernetes keeps the DBC in until every child resource is deleted, not just the restores

Reply:
There is no need for listing the restores indeed if the foreground deletion is used. Thank you for the idea, I love it how concise the code has become👍